File: src/pty.zig
const std = @import("std");
const posix = std.posix;
const testing = std.testing;
const linux = std.os.linux;

// We use extern declarations for PTY functions not in Zig's std.
extern "c" fn grantpt(fd: c_int) c_int;
extern "c" fn unlockpt(fd: c_int) c_int;
extern "c" fn ptsname(fd: c_int) ?[*:0]const u8;
extern "c" fn setsid() c_int;
extern "c" fn fork() c_int;
extern "c" fn dup2(oldfd: c_int, newfd: c_int) c_int;
extern "c" fn execvp(file: [*:0]const u8, argv: [*:null]const ?[*:0]const u8) c_int;
extern "c" fn chdir(path: [*:0]const u8) c_int;
extern "c" fn putenv(string: [*:0]u8) c_int;
extern "c" fn getenv(name: [*:0]const u8) ?[*:0]const u8;
extern "c" fn waitpid(pid: c_int, status: *c_int, options: c_int) c_int;

const TIOCSWINSZ = 0x5414;
const TIOCSCTTY = 0x540E;
const WNOHANG = 1;

const winsize = extern struct {
    ws_row: u16,
    ws_col: u16,
    ws_xpixel: u16,
    ws_ypixel: u16,
};

/// Manages a pseudo-terminal (PTY) for a child process.
/// Following Ghostty's pattern of per-surface PTY management.
pub const Pty = struct {
    master_fd: posix.fd_t,
    child_pid: c_int,

    /// Spawn a new PTY with the given dimensions, running a shell.
    pub fn spawn(
        cols: u16,
        rows: u16,
        shell: ?[]const u8,
        cwd: ?[]const u8,
    ) !Pty {
        // Open /dev/ptmx directly (equivalent to posix_openpt)
        const master_fd = posix.open(
            "/dev/ptmx",
            .{ .ACCMODE = .RDWR, .NOCTTY = true },
            0,
        ) catch return error.OpenPtyFailed;

        if (grantpt(master_fd) != 0) {
            posix.close(master_fd);
            return error.GrantPtyFailed;
        }
        if (unlockpt(master_fd) != 0) {
            posix.close(master_fd);
            return error.UnlockPtyFailed;
        }

        // Get slave name
        const slave_name = ptsname(master_fd);
        if (slave_name == null) {
            posix.close(master_fd);
            return error.PtsnameFailed;
        }

        // Set window size
        var ws = winsize{
            .ws_row = rows,
            .ws_col = cols,
            .ws_xpixel = 0,
            .ws_ypixel = 0,
        };
        _ = linux.ioctl(master_fd, TIOCSWINSZ, @intFromPtr(&ws));

        // Fork
        const pid = fork();
        if (pid < 0) {
            posix.close(master_fd);
            return error.ForkFailed;
        }

        if (pid == 0) {
            // ---- Child process ----
            _ = setsid();

            // Open slave PTY
            const slave_fd_result = posix.open(
                std.mem.span(slave_name.?),
                .{ .ACCMODE = .RDWR },
                0,
            ) catch {
                posix.exit(1);
            };

            // Set as controlling terminal
            _ = linux.ioctl(slave_fd_result, TIOCSCTTY, 0);
            _ = linux.ioctl(slave_fd_result, TIOCSWINSZ, @intFromPtr(&ws));

            // Redirect stdio
            _ = dup2(@intCast(slave_fd_result), 0);
            _ = dup2(@intCast(slave_fd_result), 1);
            _ = dup2(@intCast(slave_fd_result), 2);
            if (slave_fd_result > 2) posix.close(slave_fd_result);
            posix.close(master_fd);

            // Change directory
            if (cwd) |dir| {
                var dir_buf: [4096]u8 = undefined;
                const len = @min(dir.len, dir_buf.len - 1);
                @memcpy(dir_buf[0..len], dir[0..len]);
                dir_buf[len] = 0;
                _ = chdir(@ptrCast(dir_buf[0..len :0]));
            }

            // Set TERM environment variable
            _ = putenv(@constCast(@as([*:0]u8, @ptrCast(@constCast("TERM=xterm-256color")))));
            _ = putenv(@constCast(@as([*:0]u8, @ptrCast(@constCast("COLORTERM=truecolor")))));

            // Determine shell
            var shell_buf: [256]u8 = undefined;
            var shell_path: [*:0]const u8 = "/bin/bash";

            if (shell) |s| {
                const slen = @min(s.len, shell_buf.len - 1);
                @memcpy(shell_buf[0..slen], s[0..slen]);
                shell_buf[slen] = 0;
                shell_path = @ptrCast(shell_buf[0..slen :0]);
            } else {
                const env_shell = getenv("SHELL");
                if (env_shell) |es| shell_path = es;
            }

            const args = [_:null]?[*:0]const u8{shell_path};
            _ = execvp(shell_path, &args);
            posix.exit(1);
        }

        // ---- Parent process ----
        // Set master to non-blocking (O_NONBLOCK = 0x800 on Linux x86_64)
        const O_NONBLOCK: u32 = 0x800;
        const current_flags = linux.fcntl(master_fd, linux.F.GETFL, @as(u32, 0));
        _ = linux.fcntl(master_fd, linux.F.SETFL, current_flags | O_NONBLOCK);

        return .{
            .master_fd = master_fd,
            .child_pid = pid,
        };
    }

    /// Write data to the PTY master (sends to child's stdin).
    pub fn write(self: *const Pty, data: []const u8) !usize {
        const result = linux.write(self.master_fd, data.ptr, data.len);
        const signed: isize = @bitCast(result);
        if (signed < 0) return error.WriteFailed;
        return @intCast(result);
    }

    /// Read data from the PTY master (receives child's stdout).
    /// Returns the number of bytes read, or null if nothing available.
    pub fn read(self: *const Pty, buf: []u8) ?usize {
        const result = linux.read(self.master_fd, buf.ptr, buf.len);
        const signed: isize = @bitCast(result);
        if (signed <= 0) return null;
        return @intCast(result);
    }

    /// Check if the child process is still alive.
    pub fn isAlive(self: *const Pty) bool {
        var status: c_int = 0;
        const result = waitpid(self.child_pid, &status, WNOHANG);
        return result == 0;
    }

    /// Get the child process PID.
    pub fn childPid(self: *const Pty) u32 {
        return @intCast(self.child_pid);
    }

    /// Resize the PTY.
    pub fn doResize(self: *const Pty, cols_new: u16, rows_new: u16) void {
        var ws = winsize{
            .ws_row = rows_new,
            .ws_col = cols_new,
            .ws_xpixel = 0,
            .ws_ypixel = 0,
        };
        _ = linux.ioctl(self.master_fd, TIOCSWINSZ, @intFromPtr(&ws));
    }

    /// Close the PTY master fd.
    pub fn close(self: *Pty) void {
        posix.close(self.master_fd);
        self.master_fd = -1;
    }
};

test "pty types compile" {
    // Basic compile-time check — actually spawning a PTY requires a real system
    _ = Pty{
        .master_fd = -1,
        .child_pid = -1,
    };
}

File: src/pty.rs
use std::ffi::CString;
use std::io;
use std::os::fd::{AsRawFd, FromRawFd, OwnedFd};

/// Manages a pseudo-terminal (PTY) for a child process
pub struct Pty {
    master: OwnedFd,
    child_pid: u32,
}

impl Pty {
    /// Spawn a new PTY with the given dimensions, running a shell
    pub fn spawn(
        cols: u16,
        rows: u16,
        shell: Option<&str>,
        cwd: Option<&str>,
    ) -> io::Result<Self> {
        // Open a PTY master/slave pair
        let result = unsafe { libc::posix_openpt(libc::O_RDWR | libc::O_NOCTTY) };
        if result < 0 {
            return Err(io::Error::last_os_error());
        }
        let master_fd = unsafe { OwnedFd::from_raw_fd(result) };

        // Grant access and unlock
        if unsafe { libc::grantpt(master_fd.as_raw_fd()) } != 0 {
            return Err(io::Error::last_os_error());
        }
        if unsafe { libc::unlockpt(master_fd.as_raw_fd()) } != 0 {
            return Err(io::Error::last_os_error());
        }

        // Get slave name
        let slave_name = unsafe {
            let ptr = libc::ptsname(master_fd.as_raw_fd());
            if ptr.is_null() {
                return Err(io::Error::last_os_error());
            }
            std::ffi::CStr::from_ptr(ptr).to_owned()
        };

        // Set window size on master
        let winsize = libc::winsize {
            ws_row: rows,
            ws_col: cols,
            ws_xpixel: 0,
            ws_ypixel: 0,
        };
        unsafe {
            libc::ioctl(master_fd.as_raw_fd(), libc::TIOCSWINSZ, &winsize);
        }

        // Fork
        let pid = unsafe { libc::fork() };
        if pid < 0 {
            return Err(io::Error::last_os_error());
        }

        if pid == 0 {
            // Child process
            unsafe {
                // Create new session
                libc::setsid();

                // Open slave PTY
                let slave_fd = libc::open(slave_name.as_ptr(), libc::O_RDWR);
                if slave_fd < 0 {
                    libc::_exit(1);
                }

                // Set as controlling terminal
                libc::ioctl(slave_fd, libc::TIOCSCTTY as libc::c_ulong, 0);

                // Set window size on slave too
                libc::ioctl(slave_fd, libc::TIOCSWINSZ, &winsize);

                // Redirect stdin/stdout/stderr to slave PTY
                libc::dup2(slave_fd, 0);
                libc::dup2(slave_fd, 1);
                libc::dup2(slave_fd, 2);

                if slave_fd > 2 {
                    libc::close(slave_fd);
                }

                // Close the master fd in child
                // (OwnedFd will be dropped, but we explicitly close to be safe)
                libc::close(master_fd.as_raw_fd());

                // Change directory if requested
                if let Some(dir) = cwd {
                    if let Ok(dir_c) = CString::new(dir) {
                        libc::chdir(dir_c.as_ptr());
                    }
                }

                // Set TERM environment variable
                let term_env = CString::new("TERM=xterm-256color").unwrap();
                libc::putenv(term_env.as_ptr() as *mut _);

                // Set COLORTERM
                let colorterm = CString::new("COLORTERM=truecolor").unwrap();
                libc::putenv(colorterm.as_ptr() as *mut _);

                // Exec the shell
                let shell_path = shell
                    .map(|s| s.to_string())
                    .or_else(|| std::env::var("SHELL").ok())
                    .unwrap_or_else(|| "/bin/bash".to_string());

                let shell_c = CString::new(shell_path.as_str()).unwrap();
                let shell_name = CString::new(
                    std::path::Path::new(&shell_path)
                        .file_name()
                        .unwrap_or_default()
                        .to_str()
                        .unwrap_or("sh"),
                )
                .unwrap();

                // Login shell (prefix with -)
                let login_name = CString::new(format!(
                    "-{}",
                    shell_name.to_str().unwrap_or("sh")
                ))
                .unwrap();

                let args = [login_name.as_ptr(), std::ptr::null()];
                libc::execvp(shell_c.as_ptr(), args.as_ptr());

                // If exec fails
                libc::_exit(1);
            }
        }

        // Parent process
        // Set master to non-blocking
        unsafe {
            let flags = libc::fcntl(master_fd.as_raw_fd(), libc::F_GETFL);
            libc::fcntl(
                master_fd.as_raw_fd(),
                libc::F_SETFL,
                flags | libc::O_NONBLOCK,
            );
        }

        // Forget the OwnedFd for the master since we'll manage it ourselves
        // Actually, we keep the OwnedFd and it will close on drop
        Ok(Self {
            master: master_fd,
            child_pid: pid as u32,
        })
    }

    /// Write data to the PTY master (sends to child's stdin)
    pub fn write(&self, data: &[u8]) -> io::Result<usize> {
        let result = unsafe {
            libc::write(
                self.master.as_raw_fd(),
                data.as_ptr() as *const _,
                data.len(),
            )
        };
        if result < 0 {
            Err(io::Error::last_os_error())
        } else {
            Ok(result as usize)
        }
    }

    /// Read data from the PTY master (receives child's stdout)
    pub fn read(&self) -> Option<Vec<u8>> {
        let mut buf = vec![0u8; 65536];
        let result = unsafe {
            libc::read(
                self.master.as_raw_fd(),
                buf.as_mut_ptr() as *mut _,
                buf.len(),
            )
        };
        if result > 0 {
            buf.truncate(result as usize);
            Some(buf)
        } else {
            None
        }
    }

    /// Check if the child process is still alive
    pub fn is_alive(&self) -> bool {
        let mut status: libc::c_int = 0;
        let result = unsafe { libc::waitpid(self.child_pid as i32, &mut status, libc::WNOHANG) };
        // waitpid returns 0 if child is still running, >0 if exited, -1 on error
        result == 0
    }

    /// Get the child process PID
    pub fn child_pid(&self) -> u32 {
        self.child_pid
    }

    /// Resize the PTY
    pub fn resize(&self, cols: u16, rows: u16) {
        let winsize = libc::winsize {
            ws_row: rows,
            ws_col: cols,
            ws_xpixel: 0,
            ws_ypixel: 0,
        };
        unsafe {
            libc::ioctl(self.master.as_raw_fd(), libc::TIOCSWINSZ, &winsize);
        }
    }
}
